### PR TITLE
Do not truncate Logger timestamps

### PR DESCRIPTION
**Pull request by binarylogic**
Is there any reason why the precision on `Logger` timestamps is being truncated? If so, can this be commented / documented somewhere?

A few reasons we should _not_ do this:

1. We're seeing log messages share the same timestamp, thus presenting them out of order in external logging systems.
2. The syslog specification allows up to a microsecond precision: https://tools.ietf.org/html/rfc5424
3. The backend should be solely responsible for removing precision.

Thanks for taking a look!

### CODE REVIEW

File: lib/logger/lib/logger/utils.ex [hunk 1]
```diff
@@ -244,14 +244,14 @@ defmodule Logger.Utils do
...
       end
-    {date, {hours, minutes, seconds, div(micro, 1000)}}
+    {date, {hours, minutes, seconds, micro}}
   end
 
   @doc """
   Formats time as chardata.
   """
   def format_time({hh, mi, ss, ms}) do
-    [pad2(hh), ?:, pad2(mi), ?:, pad2(ss), ?., pad3(ms)]
```
Comment (josevalim) [inline lib/logger/lib/logger/utils.ex]: In order to not change the output currently logged, let's make this be:

```elixir
[pad2(hh), ?:, pad2(mi), ?:, pad2(ss), ?., ms |> div(1000) |> pad3()]
```

If you want to log with higher precision, then you can always call your own `format_time` function.

This way you get higher precision in the backend but remove the precision only when formatting, as you mentioned.

Comment (binarylogic) [inline lib/logger/lib/logger/utils.ex]: Sure, makes sense. I'll update.

Comment (binarylogic) [inline lib/logger/lib/logger/utils.ex]: Ok, all set.

**Comment by fishcakez** [comment]
This would be a breaking change, so would have to be opt-in. I would rather we support an option to use a date time struct.

**Comment by josevalim** [comment]
@fishcakez you are right. We are changing the meaning of the tuple element. I think having an option for using time structs is likely the way to go here.

**Comment by binarylogic** [comment]
Sure, do you mind explaining why this is breaking? And does this mean Elixir is locked in at millisecond precision for logging, by default?

**Comment by fishcakez** [comment]
> do you mind explaining why this is breaking?

We support custom formatting functions, and these receive the timestamp tuple, which is currently milliseconds. See `Logger.Formatter.compile({module, function})`. Therefore we can't change the timestamp to be microseconds, instead of milliseconds.

> And does this mean Elixir is locked in at millisecond precision for logging, by default?

Yes

**Comment by binarylogic** [comment]
Thanks. I'll update this to take an option.

**Comment by josevalim** [comment]
Hi @binarylogic! I am closing this PR as, per the discussion above, it is backwards incompatible. Instead, we should probably move to a NaiveDateTime struct, configured via an option. Thank you!

**Pull request closed without merging**